Alex and Happy Part 2/article.md

```diff
@@ -0,0 +1,1323 @@
+1 + <b><i>num</i> + exp</b>
+1 + <b>2</b> + <i>exp</i>
+1 + 2 + <b><i>num</i></b>
+1 + 2 + <b>3</b>
+</pre>
+
+n.b.: **Bold** means the result of the derivation, and <i>italic</i> means the derivation that will be expanded next.
+
+Whilst an LR parser would do the following:
+
+<pre>
+<i>exp</i>
+<b>num + <i>exp</i></b>
+num + <b>num + <i>exp</i></b>
+num + num + <b><i>num</i></b>
+num + <i>num</i> + <b>3</b>
+<i>num</i> + <b>2</b> + 3
+<b>1</b> + 2 + 3
+</pre>
+
+For more information on leftmost and rightmost derivations, see [LL and LR Parsing Demystified](https://blog.reverberate.org/2013/07/ll-and-lr-parsing-demystified.html).
+
+Another difference is that Alex is a lexer generator and Happy a parser generator.
+It means that instead of writing Haskell files (`.hs`) like you do with Megaparsec, you will write Alex (`.x`) and Happy (`.y`) files, which in turn will automatically generate Haskell files with implementations of the lexer and parser recognizing the described grammar.
+
+Below are some pros and cons of Happy in comparison to Megaparsec:
+
+Pros:
+* LR parsers are more powerful than LL parsers, allowing to parse more complex grammars with greater ease.
+* Ambiguous grammars are reported to the user, allowing a greater trust in the parser's output and saving the debugging headache.
+* LR parsers can handle left-recursion, while LL parsers can't.
+* The performance will be more predictable, as Megaparsec can cause exponential time complexity with the use of `try`. In addition, Happy's "static" generated LALR algorithm has better known complexity constraints.
+
+Cons:
+* Less idiomatic, as you need to write Alex and Happy grammars instead of Haskell code.
+* Less flexible since you can't touch the algorithm on how Happy works. Meanwhile, you can easily create your functions for parser combinators.
+
+Comparing the implementations of both parsing techniques is beyond the scope of this post. Still, if you are interested, you may read the parser combinators article linked above.
+
+## How it works
+
+An in-depth explanation of how an LR(1) parser works is well beyond the scope of this tutorial, but for this, we recommend you read our article, [How to Implement an LR(1) Parser](https://serokell.io/blog/how-to-implement-lr1-parser).
+We will, however, introduce a few basic concepts required to understand this article better.
+
+### Terminals, non-terminals, and production rules
+
+A **terminal** is a lexical item of the language; in our case they will be tokens that were produced by Alex.
+
+A **non-terminal** is an arbitrary identifier that's replaced by other terminals and non-terminals (terminals and non-terminals together are called **symbols**) to produce derivations.
+In the example below, `myNonTerminal` is non-terminal.
+
+A **production rule** is a rule describing how to "produce" a string that is a sentence corresponding to the grammar.
+It has form `a : b { action }`, which looks like so in Happy:
+
+```happy
```

Comment:
I had some difficulties grasping what this is.

You later have nice examples such as dec, I think it will be much clearer with those (perhaps obscure how actions look :) ) .

Reply:
I made a small concrete example, I hope it will be easier to understand with it.